src/ModelContextProtocol.AspNetCore/StatefulSessionManager.cs

```diff
@@ -0,0 +1,242 @@
+                    return;
+                }
+            }
+        }
+        catch
+        {
+            await newSession.DisposeAsync();
+            throw;
+        }
+        finally
+        {
+            _idlePruningLock.Release();
+        }
+    }
+
+    /// <summary>
+    /// Performs a single pass of idle session pruning, removing sessions that exceed the idle timeout
+    /// or when the maximum idle session count is exceeded.
+    /// </summary>
+    public async Task PruneIdleSessionsAsync(CancellationToken cancellationToken)
+    {
+        await _idlePruningLock.WaitAsync(cancellationToken);
+        try
+        {
+            PruneIdleSessionsUnsynchronized();
+        }
+        finally
+        {
+            _idlePruningLock.Release();
+        }
+    }
+
+    private void PruneIdleSessionsUnsynchronized()
+    {
+        var idleActivityCutoff = _idleTimeoutTicks switch
+        {
+            < 0 => long.MinValue,
+            var ticks => _timeProvider.GetTimestamp() - ticks,
+        };
+
+        // We clear the lists at the start of pruning rather than the end so we can use them between runs
+        // to find the most idle sessions to remove one-at-a-time if necessary to make room for new sessions.
+        _idleTimeStamps.Clear();
+        _idleSessionIds.Clear();
+        _nextIndexToPrune = -1;
+
+        foreach (var (_, session) in this)
+        {
+            if (session.IsActive || session.SessionClosed.IsCancellationRequested)
+            {
+                // There's a request currently active or the session is already being closed.
+                continue;
+            }
+
+            if (session.LastActivityTicks < idleActivityCutoff)
+            {
+                LogIdleSessionTimeout(session.Id, _idleTimeout);
+                RemoveAndCloseSession(session.Id);
+                continue;
+            }
+
+            // Add the timestamp and the session
+            _idleTimeStamps.Add(session.LastActivityTicks);
+            _idleSessionIds.Add(session.Id);
+        }
+
+        if (_idleTimeStamps.Count > _maxIdleSessionCount)
+        {
+            // Sort only if the maximum is breached and sort solely by the timestamp.
+            EnsureIdleSessionsSortedUnsynchronized();
+
+            var sessionsToPrune = CollectionsMarshal.AsSpan(_idleSessionIds)[..^_maxIdleSessionCount];
+            foreach (var id in sessionsToPrune)
+            {
+                LogIdleSessionLimit(id, _maxIdleSessionCount);
+                RemoveAndCloseSession(id);
+            }
+            _nextIndexToPrune = _maxIdleSessionCount;
+        }
+    }
+
+    private void EnsureIdleSessionsSortedUnsynchronized()
+    {
+        if (_nextIndexToPrune > -1)
+        {
+            // Already sorted.
+            return;
+        }
+
+        var timestamps = CollectionsMarshal.AsSpan(_idleTimeStamps);
+        timestamps.Sort(CollectionsMarshal.AsSpan(_idleSessionIds));
```

Comment:
In a situation where we're at our limit and so every new request is going to need to prune something, are we going to end up doing an O(n log n) sort each time? Is that ok because the length is small enough, or because it should generally be sorted already and won't actually be O(n log n), or something else?

Reply:
> or because it should generally be sorted already and won't actually be O(n log n)

Basically, if there are a lot of idle sessions to sort, we should have a lot of idle sessions to iterate through before we're forced to repopulate the list and sort again. Technically the big-O worst-case asymptotic complexity probably is O(n log n) when we're sitting at exactly the limit of idle sessions and there's only one request every 5 seconds which makes the background loop force a resort, but the average case should be closer to O(1).

Reply:
> should be closer to O(1)

You mean O(n), right?